mlflow/pipelines/steps/train.py

```diff
@@ -166,13 +193,111 @@ def _run(self, output_directory):
 
         return card
 
+    def _get_leaderboard_df(self, run, eval_metrics):
+        import pandas as pd
+
+        mlflow_client = MlflowClient()
+        exp_id = _get_experiment_id()
+
+        primary_metric_greater_is_better = self.evaluation_metrics[
+            self.primary_metric
+        ].greater_is_better
+        primary_metric_order = "DESC" if primary_metric_greater_is_better else "ASC"
+
+        search_max_results = 100
+        search_result = mlflow_client.search_runs(
+            experiment_ids=exp_id,
+            run_view_type=ViewType.ACTIVE_ONLY,
+            max_results=search_max_results,
+            order_by=[f"metrics.{self.primary_metric}_on_data_validation {primary_metric_order}"],
+        )
+
+        metric_names = self.evaluation_metrics.keys()
+        metric_keys = [f"{metric_name}_on_data_validation" for metric_name in metric_names]
+
+        leaderboard_items = []
+        for old_run in search_result:
+            if all(metric_key in old_run.data.metrics for metric_key in metric_keys):
+                leaderboard_items.append(
+                    {
+                        "Run ID": old_run.info.run_id,
+                        "Run Time": datetime.datetime.fromtimestamp(
+                            old_run.info.start_time // 1000
+                        ),
+                        **{
+                            metric_name: old_run.data.metrics[metric_key]
+                            for metric_name, metric_key in zip(metric_names, metric_keys)
+                        },
+                    }
+                )
+
+        top_leaderboard_items = [
+            {"Model Rank": i + 1, **t} for i, t in enumerate(leaderboard_items[:2])
+        ]
+
+        if (
+            len(top_leaderboard_items) == 2
+            and top_leaderboard_items[0][self.primary_metric]
+            == top_leaderboard_items[1][self.primary_metric]
+        ):
+            # If top1 and top2 model primary metrics are equal,
+            # then their rank are both 1.
+            top_leaderboard_items[1]["Model Rank"] = "1"
+
+        top_leaderboard_item_index_values = ["Best", "2nd Best"][: len(top_leaderboard_items)]
+
+        latest_model_item = {
+            "Run ID": run.info.run_id,
+            "Run Time": datetime.datetime.fromtimestamp(run.info.start_time // 1000),
+            **eval_metrics["validation"],
+        }
+
+        for i, leaderboard_item in enumerate(leaderboard_items):
+            latest_value = latest_model_item[self.primary_metric]
+            historical_value = leaderboard_item[self.primary_metric]
+            if (primary_metric_greater_is_better and latest_value >= historical_value) or (
+                not primary_metric_greater_is_better and latest_value <= historical_value
+            ):
+                latest_model_item["Model Rank"] = str(i + 1)
+                break
+        else:
+            latest_model_item["Model Rank"] = f"> {len(leaderboard_items)}"
```

Comment:
```suggestion
        for i, leaderboard_item in enumerate(leaderboard_items):
            if latest_model_item["Run ID"] == leaderboard_item["Run ID"]:
                latest_model_item["Model Rank"] = str(i + 1)
                break
        else:
            latest_model_item["Model Rank"] = f"> {len(leaderboard_items)}"
```

Can we just compare run IDs here?

Reply:
I don't think so.

We only query the top 100 historical runs, 
and the latest run ID might not be in the top 100 historical runs,
and even if the latest run ID might not be in the top 100 historical runs, the latest run rank might still be in the range of (0, 100).

Reply:
Makes sense!

Reply:
Makes sense now. My approach doesn't work when `leaderboard_items` contains ties:

|run_id|primary metric value|rank (current implementation)|rank (suggested implementation)|
|:-|:-|:-|:-|
|1|1|1|1
|...|...|...|...
|98|98|98|98|
|99|99|99|99|
|100|99|99|99|
|101 (latest run)|100|99|> 100|